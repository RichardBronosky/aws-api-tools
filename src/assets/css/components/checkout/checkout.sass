@import '../../mixins'
@import '../../variables'

.checkout-page
  margin-top: -2.5rem
  .checkout-page-header
    margin: 0
    hewight: 1.25rem
    h1
      line-height: 1
      font-weight: 300
      font-size: $txt-xlarge
      color: $dark-grey
      padding: 0
      margin-top: 0.8rem
      margin-bottom: 0.4rem
      +large-up
        padding-left: 1rem
    .if-have-account
      margin-top: 0.5em
      font-size: $txt-medium
      font-weight: 500
      a
        font-weight: bold
        color: $primary
        &:hover
          text-decoration: underline
    .if-not-have-account
      font-size: $txt-small
  .checkout-section-steps
    margin-top: 0 !important
    .checkout-section-container
      margin-top: 0
      +medium-down
        margin-right: 1em
  .checkout-container
    position: -webkit-sticky
    position: sticky
    top: 5.25em!important
    right: 0
    height: 100%
    z-index: 1
    margin-top: 0
    +large-up
      max-height: 33rem
    +medium-down
      border-top: solid 2px $grey
      bottom: -1px
      top: unset !important
      right: 1em
      left: 1em
      width: 100%
    +small
      margin-top: 1.5em
      font-size: $txt-small
    +landscape
      margin-top: 1.5em
      font-size: $txt-small 
    +ie
      position: static
    .bold
      font-weight: 600
