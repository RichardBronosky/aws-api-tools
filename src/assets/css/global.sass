@import 'variables'
@import 'mixins'
@import 'global/_all'

body
  overflow-x: hidden
  padding: 0px
  margin: 0px
  font-family: "Open Sans", sans-serif
  background-color: #F7F7F7
  width: 100%

.content-wrapper
  strong
    font-weight: 700
  +large-up
    padding-top: 125px
  +medium-down
    padding-top: 127px
  .generated-page div:first-of-type > .banner:first-of-type
    margin-top: -0.8rem !important
  &:focus
    outline: none

button
  cursor: pointer
  &.primary
    background-color: $primary
    color: $primary-invert
    border: 0
    text-align: center
    -webkit-box-sizing: border-box
    box-sizing: border-box
    font-size: $txt-small
    text-transform: uppercase
    margin: 0 auto
    border-radius: 0
    font-weight: bold
    margin-top: .5rem
    margin-bottom: .5rem
    padding: 0.625rem 1.625rem
    &.unavailable
      background: $grey
      padding: 0.625rem 1.625rem
      &:hover
        background: $grey
    &:hover
      background: $light-blue
    &:focus
      text-decoration: underline
svg 
  shape-rendering: auto
.scroll-to-top, .scrolled-cart
  display: none
  position: fixed
  bottom: 1rem
  right: 1rem
  background: $primary
  border-radius: 50%
  width: 3rem
  height: 3rem
  text-align: center
  z-index: 10
  opacity: 0
  -webkit-animation-duration: 0.25s
  animation-duration: 0.25s
  -webkit-animation-fill-mode: both
  animation-fill-mode: both
  box-shadow: 0 2px 5px rgba(0,0,0,0.5)
  &.cart
    bottom: auto
    top: 1rem
  &:hover
    background: $light-blue
  +landscape
    display: block
  p
    padding: 0.8rem
    color: $primary-invert
    transform: scale(2.75,1.75)


.scrolled-cart
  bottom: auto
  top: 1rem
  .icon
    &.cart
      height: 100%
      padding: 0
      width: 50%
      margin-top: 0
  .round-red
    position: absolute
    color: $primary-invert
    font-weight: bold
    background: $red
    width: 1.25rem
    height: 1.25rem
    border-radius: 50%
    top: -0.3rem
    right: -0.3rem
    p
      position: relative
      top: 0.11rem
      right: 0.02rem
      padding: 0
      color: $primary-invert
      font-size: $txt-xsmall
      transform: unset

.skip
  position: absolute
  left: -10000px
  top: auto
  width: 1px
  height: 1px
  overflow: hidden
  &:focus, &:active, &:hover
    left: 0
    top: 0
    width: auto
    height: auto
    overflow: visible
    z-index: 9999
    background-color: #fff
    padding: 1rem
    margin: 1rem
    border: 1px solid #DEDEDE
    box-shadow: 0 6px 20px #AFAFAF  


.card
  height: auto
  width: 100%
  display: block
  box-shadow: 0 2px 5px rgba(0, 0, 0, 0.2)
  background-color: $primary-invert
  img
    width: 100%
  &.pop-out
    &:focus,
    &:active,
    &:hover,
    &.focused,
    &:focus-within
      -webkit-transition: all .2s ease
      -moz-transition: all 0.2s ease
      -ms-transition: all 0.2s ease
      -o-transition: all 0.2s ease
      transition: all 0.2s ease
      transform: scale(1.01)
      box-shadow: 0 6px 20px #AFAFAF
      z-index: 9
      .buy-now-popup
        +large-up
          display: inline-block
.border-top
  border-top: 1px solid $grey
.head--megamenu
  .full-width
    left: 0
    right: 0
    margin: 0 auto

.modal-default
  position: absolute
  top: 40px
  left: 40px
  right: 40px
  bottom: 40px
  background-color: $primary-invert
.modal-overlay
  position: fixed
  z-index: 10
  top: 0
  left: 0
  right: 0
  bottom: 0
  background-color: rgba(0,0,0,0.4)
.horizontal-center
  display: flex
  justify-content: center
  img
    max-height: 8rem
  +large-up
    justify-content: right
.vertical-center
  display: flex
  flex-direction: column
  -ms-flex-direction: column
  justify-content: center

.blue-action-btn, .link-action-btn
  background-color: $primary
  color: $primary-invert
  border: 0px
  width: 100%
  padding: 15px
  text-align: center
  box-sizing: border-box
  font-size: $txt-medium
  text-transform: uppercase
  border-radius: 0
  &:hover
    background-color: $light-blue
  +focus-508
  &.checkout
    display: block

.link-action-btn
  background-color: inherit 
  color: inherit 
  &:hover
    text-decoration: underline
    background-color: inherit
    color: $primary 
  +focus-508
  &.checkout
    display: block




.seo-page-heading
  background-color: $primary
  margin-top: 13px
  +large-up
    margin-top: 4px

.home-heading
  height: 0
  padding: 0
  opacity: 0

.pagination
  width: 100%
  .ais-Pagination
    display: flex
    justify-content: center
    width: 100%
  .ais-Pagination-list
    display: flex
    justify-content: space-between
    width: 75%


  //Inputs



.product-in-room
  img
    +large-up

      object-fit: cover


.close
  width: 40px
  height: 40px
  border-radius: 40px
  position: relative
  z-index: 1
  cursor: pointer
  &:before
    content: '+'
    color: $dark-grey
    position: absolute
    z-index: 2
    transform: rotate(45deg)
    font-size: 50px
    line-height: 1
    top: -5px
    left: 6px
    transition:  $animation
  &:after
    content: ''
    position: absolute
    top: 0
    left: 0
    width: 100%
    height: 100%
    border-radius: 100%
    background: $red
    z-index: 1
    transition:  $animation
    transform: scale(0.01)
  &:hover:after
    transform: scale(1)
  &:hover:before
    transform: scale(0.8) rotate(45deg)
    color: #fff
.head--iconbar-cart
  .close
    width: 20px
    height: 20px
    &:before
      top: -5px
      left: 2px
      font-size: 30px

.ril-inner
  background: white !important

