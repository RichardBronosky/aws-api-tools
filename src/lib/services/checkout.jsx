const axios = require('axios')

// TODO: Change to furniture.com API? Right now browser will show auth-token.
export const fetchAddressLookup = address =>
  axios
    .get(`https://api.edq.com/capture/address/v2/search?country=usa&take=10&shoppingState=&query=${ address }`, {
      headers: {
        'auth-token': '26dd9bb3-9ee2-41a7-a777-c3ec24cd411b',
      },
    })
    .then(response => response.data)

export const fetchPromotions = body =>
  axios.post(`${ process.env.GATSBY_PROMOTIONS_SERVICE_URL }/get_pricing/`, body).then(response => response.data)

/* Order Service */
export const getOrder = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/getOrder/`, body).then(response => response.data)

export const getToken = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/getToken/`, body).then(response => response.data)

export const createOrder = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/createOrder/`, body).then(response => response.data)

export const updateLineItems = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/updateLineItems/`, body).then(response => response.data)

export const updateAddress = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/updateAddress/`, body).then(response => response.data)

export const updateDelivery = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/updateDelivery/`, body).then(response => response.data)

export const updatePayment = body =>
  axios.post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/updatePayment/`, body).then(response => response.data)

export const placeOrder = (orderId, session, TL_RTG) =>
  axios
    .post(`${ process.env.GATSBY_ORDER_SERVICE_URL }/placeOrder/`, { orderId: orderId, session: session, TL_RTG: TL_RTG })
    .then(response => response.data)
