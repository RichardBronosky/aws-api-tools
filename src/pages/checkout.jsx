import React from 'react'
import Helmet from 'react-helmet'
import Layout from '../components/layout'
import CheckoutMain from '../components/checkout'

export default ({ props }) => (
  <Layout { ...props } checkout>
    <Helmet title={ 'Checkout - Rooms To Go' } meta={ [{ name: 'robots', content: 'noindex, nofollow' }] } />
    <CheckoutMain />
  </Layout>
)
