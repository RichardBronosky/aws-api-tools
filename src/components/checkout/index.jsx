import React from 'react'
import { connect } from 'react-redux'
import Modal from 'react-modal'
import '../../assets/css/components/checkout/checkout.sass'
import loaderDark from '../../assets/images/loader-dark.svg'
import { getFromBrowserStorage } from '../../lib/helpers/storage'

import CondensedContact from '../shared/condensed-contact-links'
import CheckoutSticky from './checkout-sticky/checkout-sticky'
import CheckoutStep from './checkout-parts/checkout-step'
import CheckoutEmpty from './checkout-parts/checkout-empty'

import ShippingSection from './checkout-parts/shipping-section'
import DeliverySection from './checkout-parts/delivery-section'
import PaymentSection from './checkout-parts/payment-section'
import ReviewSection from './checkout-parts/review-section'
import { setupCheckoutAnalytics } from '../../lib/helpers/google-tag-manager'
import DeclineModal from './checkout-parts/decline-modal'
import { setCart } from '../../redux/modules/cart'
import { setOrder, setCheckoutStep, setCheckoutStepsCompleted } from '../../redux/modules/checkout'
import { createOrder, updateLineItems } from '../../lib/services/checkout'
import { getLineItems } from '../../lib/helpers/checkout/global'
import { getCurrentLocation } from '../../lib/helpers/geo-location'

if (process.env.NODE_ENV !== 'test') Modal.setAppElement('#___gatsby')

class Checkout extends React.Component {
  state = {
    paymentInfo: {
      rtgCreditInfo: {
        cardNumber: '',
        zip: '',
      },
      financePlan: {
        code: '',
        hasPayments: true,
      },
      useGiftCard: false,
      giftCardNumber: '',
      giftCardPin: '',
      giftCardInfo: [],
      addAnotherCard: false,
      billingDifferent: false,
      billingSubmitted: false,
    },
    paymentInvalidFields: [],
    reviewInfo: {
      acceptTerms: false,
      acceptPickupTerms: false,
    },
    reviewInvalidFields: [],
    declineModalOpen: false,
    declineCloseLoading: false,
    declineType: 'credit card',
  }

  async componentDidMount() {
    const {
      order,
      onSetOrder,
      checkoutStep,
      onSetCheckoutStep,
      checkoutStepsCompleted,
      onSetCheckoutStepsCompleted,
      onSetCart,
    } = this.props
    const cartStorage = getFromBrowserStorage('local', 'cart')
    if (cartStorage) {
      onSetCart(cartStorage)
      setupCheckoutAnalytics(cartStorage)
    }
    const orderStorage = getFromBrowserStorage('session', 'order')
    const hasDeprecatedPaymentType = this.checkPaymentType(orderStorage)
    if (
      orderStorage &&
      orderStorage.lineItems &&
      orderStorage.lineItems.length > 0 &&
      JSON.stringify(order) !== JSON.stringify(orderStorage) &&
      !hasDeprecatedPaymentType
    ) {
      onSetOrder(orderStorage)
    } else if (hasDeprecatedPaymentType || !orderStorage || (orderStorage && !orderStorage.orderId)) {
      const location = getCurrentLocation()
      const lineItems = getLineItems()
      const orderResp = await createOrder({
        lineItems: lineItems,
        region: location.region,
        zone: parseInt(location.price_zone),
        distribution_index: parseInt(location.distribution_index),
      })
      onSetCheckoutStep('shipping')
      if (orderResp && orderResp.orderId) {
        onSetOrder(orderResp)
      }
    } else if (orderStorage && orderStorage.orderId) {
      const lineItems = getLineItems()
      const orderResp = await updateLineItems({
        orderId: orderStorage.orderId,
        lineItems: lineItems,
      })
      if (orderResp && orderResp.orderId) {
        onSetOrder(orderResp)
      }
    }
    const checkoutStepStorage = getFromBrowserStorage('session', 'checkoutStep')
    if (order && checkoutStepStorage && checkoutStep !== checkoutStepStorage) {
      onSetCheckoutStep(checkoutStepStorage)
    }
    const checkoutStepsCompletedStorage = getFromBrowserStorage('session', 'checkoutStepsCompleted')
    if (order && checkoutStepsCompletedStorage && checkoutStepsCompleted !== checkoutStepsCompletedStorage) {
      onSetCheckoutStepsCompleted(checkoutStepsCompletedStorage)
    }
  }

  checkPaymentType(order) {
    if (order && order.paymentInfo) {
      return order.paymentInfo.some(payment => payment.paymentType === 'CYBER')
    }
    return false
  }

  render() {
    const { order, isMobile, declineModalInfo } = this.props
    return (
      <>
        { order && order.lineItems && order.lineItems.length > 0 && (
          <section className="cell small-12 grid-x grid-margin-y checkout-page">
            <div className="cell small-12 checkout-page-header">
              <h1>CHECKOUT</h1>
            </div>
            <div
              className={ `cell small-12 grid-x grid-margin-y checkout-section-steps ${
                !this.props.isMobile ? 'grid-padding-x' : ''
              }` }
            >
              <div className="checkout-section-container cell small-12 large-9 grid-x grid-margin-y">
                <div id="shipping" className="cell small-12">
                  <CheckoutStep
                    sectionTitle="Shipping Address"
                    sectionType="shipping"
                    sectionNumber="1"
                    nextSection="delivery"
                  >
                    <ShippingSection />
                  </CheckoutStep>
                </div>
                <div id="delivery" className="cell small-12">
                  <CheckoutStep
                    sectionTitle="Delivery"
                    sectionType="delivery"
                    sectionNumber="2"
                    previousSection="shipping"
                    nextSection="payment"
                  >
                    <DeliverySection />
                  </CheckoutStep>
                </div>
                <div id="payment" className="cell small-12">
                  <CheckoutStep
                    sectionTitle="Payment"
                    sectionType="payment"
                    sectionNumber="3"
                    previousSection="delivery"
                    nextSection="review"
                  >
                    <PaymentSection />
                  </CheckoutStep>
                </div>
                <div id="review" className="cell small-12">
                  <CheckoutStep
                    sectionTitle="Review Order"
                    sectionType="review"
                    sectionNumber="4"
                    previousSection="payment"
                  >
                    <ReviewSection />
                  </CheckoutStep>
                </div>
              </div>
              <div
                className={ `checkout-container cell small-12 large-3 ${ !this.props.isMobile ? 'grid-margin-y' : '' }` }
              >
                <CheckoutSticky
                  isCheckout
                  tax={ order.tax }
                  deliveryCost={ order.totalDeliveryCharge }
                  isMobile={ isMobile }
                />
              </div>
              <div className="cell small-12 large-9 checkout-contact">
                <CondensedContact />
              </div>
            </div>
          </section>
        ) }
        { order && order.lineItems && order.lineItems.length < 1 && <CheckoutEmpty /> }
        { !order && <img alt="checkout loading" src={ loaderDark } /> }
        <DeclineModal
          modalOpen={ declineModalInfo.declineModalOpen }
          type={ declineModalInfo.declineType }
          loading={ declineModalInfo.declineCloseLoading }
        />
      </>
    )
  }
}

const mapStateToProps = state => {
  return {
    ...state.global,
    ...state.cart,
    order: state.checkout.order,
    checkoutStep: state.checkout.checkoutStep,
    checkoutStepsCompleted: state.checkout.checkoutStepsCompleted,
    declineModalInfo: state.checkout.declineModalInfo,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onSetCart: cart => dispatch(setCart(cart)),
    onSetOrder: order => dispatch(setOrder(order)),
    onSetCheckoutStep: checkoutStep => dispatch(setCheckoutStep(checkoutStep)),
    onSetCheckoutStepsCompleted: checkoutStepsCompleted => dispatch(setCheckoutStepsCompleted(checkoutStepsCompleted)),
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Checkout)
