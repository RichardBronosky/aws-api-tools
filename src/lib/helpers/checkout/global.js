import * as Sentry from '@sentry/browser'
import { store } from '../../../redux/store'
import {
  setOrder,
  setCheckoutStep as setCheckoutStepRedux,
  setCheckoutStepLoading,
  initialState,
  setShippingInvalidFields,
  setDeclineModalInfo,
} from '../../../redux/modules/checkout'
import { validateShippingStep, setAddress } from './shipping-section'
import { validateDeliveryStep, canSkipDelivery } from './delivery-section'
import { validatePaymentStep } from './payment-section/payment-section'
import { redactCCFromOrder } from './payment-section/credit-card'
import { validateReviewStep, placeOrderReview } from './review-section'
import { checkoutStepAnalytics } from '../google-tag-manager'
import { getRegionZone } from '../geo-location'
import { getCart } from '../cart'
import { fetchAddressLookup } from '../../services/checkout'
import { submitBillingAddress } from './payment-section/billing-address'

export const getOrder = () => {
  return store.getState().checkout.order
}

export const setOrderInfo = (info, field) => {
  const order = getOrder()
  const newOrder = {
    ...order,
    [field]: info,
  }
  store.dispatch(setOrder(newOrder))
}

export const setCheckoutStep = async (e, checkoutStep, nextStep, placeOrder = false) => {
  e && e.preventDefault()
  store.dispatch(setCheckoutStepLoading(true))
  const invalidFields = await validateCheckoutStep(checkoutStep, nextStep)
  if (!invalidFields || invalidFields.length < 1) {
    if (placeOrder) {
      await placeOrderReview()
    } else {
      if (checkoutStep === 'shipping' && nextStep === 'delivery') {
        const canSkip = await canSkipDelivery()
        if (canSkip) {
          store.dispatch(setCheckoutStepRedux('payment'))
          checkoutStepAnalytics('payment', canSkip)
        } else {
          store.dispatch(setCheckoutStepRedux('delivery'))
          checkoutStepAnalytics('delivery')
        }
      } else {
        store.dispatch(setCheckoutStepRedux(nextStep))
        checkoutStepAnalytics(nextStep)
      }
    }
    if (checkoutStep !== nextStep) {
      if (typeof window !== 'undefined') {
        let checkoutStepEl = document.querySelector('#' + nextStep)
        if (checkoutStepEl) {
          let topOfElement = checkoutStepEl.offsetTop - 85
          try {
            window.scroll({ top: topOfElement, behavior: 'smooth', block: 'start' })
          } catch {
            //do nothing
          }
        }
      }
    }
  }
  store.dispatch(setCheckoutStepLoading(false))
}

export const validateCheckoutStep = (checkoutStep, nextStep) => {
  if (checkoutStep === 'shipping') {
    return validateShippingStep(nextStep)
  }
  if (checkoutStep === 'delivery') {
    return validateDeliveryStep(nextStep)
  }
  if (checkoutStep === 'payment') {
    return validatePaymentStep(nextStep)
  }
  if (checkoutStep === 'review' && !nextStep) {
    return validateReviewStep()
  }
}

export const reportToSentry = (err, order, callName, eventLabel) => {
  Sentry.configureScope(scope => {
    try {
      scope.setExtra('order', redactCCFromOrder(order))(err && !err.response) &&
        scope.setExtra('err', err)(err && err.response) &&
        scope.setExtra('API error response', err.response)
    } catch (e) {}
  })
  Sentry.captureException(`Checkout Error (${ callName } ${ eventLabel })`)
}

export const getRemainingTotal = (
  deliveryComplete = true,
  prevTotal = null,
  isCheckout = false,
  applyFinance = true,
  shippingComplete = true
) => {
  const order = getOrder()
  let total = 0
  let appliedGiftCards = 0
  let appliedFinance = 0
  if (prevTotal) {
    total = prevTotal
  }
  if (order && order.total && shippingComplete && isCheckout) {
    total = order.total
    if (!deliveryComplete && order.tax) {
      total -= order.tax
    }
    if (deliveryComplete) {
      if (order.paymentInfo && order.paymentInfo.length > 0) {
        const giftCards = order.paymentInfo.filter(payment => payment.paymentType === 'GIFT')
        if (
          giftCards.length > 0 &&
          giftCards[0].authorizedAmount > 0 &&
          order.giftCardInfo &&
          order.giftCardInfo.useGiftCard
        ) {
          for (let i = 0, n = giftCards.length; i < n; i++) {
            appliedGiftCards = appliedGiftCards + giftCards[i].authorizedAmount
          }
          total = total - appliedGiftCards
        }
        const rtgFin = order.paymentInfo.filter(payment => payment.paymentType === 'FIN')
        if (rtgFin.length > 0 && applyFinance) {
          appliedFinance = rtgFin[0].authorizedAmount
          total = total - appliedFinance
        }
      }
    }
  }
  return {
    total: total,
    gift: appliedGiftCards,
    fin: appliedFinance,
    tax: order && order.tax,
    deliveryCost: order && order.totalDeliveryCharge,
    totalSavings: order && order.promotions && order.promotions.totalSavings,
  }
}

export const clearCheckoutState = (clearOrderId = false) => {
  const order = getOrder()
  let newOrder = {
    ...initialState.order,
    shippingAddress: {
      ...order.shippingAddress,
    },
    contact: {
      ...order.contact,
    },
    lineItems: [],
    total: 0,
    tax: 0,
    totalDeliveryCharge: 0,
  }
  if (!clearOrderId) {
    newOrder.orderId = order.orderId
  } else {
    newOrder.orderId = null
  }
  store.dispatch(setOrder(newOrder))
  store.dispatch(setCheckoutStepRedux('shipping'))
}

export const setDeclineModalClose = () => {
  store.dispatch(
    setDeclineModalInfo({
      ...initialState.declineModalInfo,
    })
  )
}

export const getLineItems = () => {
  const cart = getCart()
  let lineItems = []
  if (cart && cart.cartItems) {
    for (let i = 0, n = cart.cartItems.length; i < n; i++) {
      const index = lineItems.findIndex(product => product.sku === cart.cartItems[i].sku)
      if (index > -1) {
        lineItems[index].quantity = lineItems[index].quantity + cart.cartItems[i].quantity
      } else {
        lineItems.push({
          sku: cart.cartItems[i].sku,
          quantity: parseInt(cart.cartItems[i].quantity),
          childItems: getProductChildren(cart.cartItems[i]),
          deliveryType: cart.cartItems[i].product.delivery_type,
          title: cart.cartItems[i].product.title,
          category: cart.cartItems[i].product.category,
          unitPrice: cart.cartItems[i].product.unitPrice,
          addons:
            cart.cartItems[i].activeAddons &&
            cart.cartItems[i].activeAddons.map(addon => ({
              sku: addon.sku,
              quantity: addon.quantity,
              required:
                cart.cartItems[i].product.addon_items &&
                cart.cartItems[i].product.addon_items.filter(addon_item => addon_item.sku === addon.sku)[0]
                  .addon_required,
            })),
        })
      }
      if (
        !cart.cartItems[i].activeAddons ||
        (cart.cartItems[i].activeAddons && cart.cartItems[i].activeAddons.length > 0)
      ) {
        const requiredAddons =
          cart.cartItems[i].product.addon_items &&
          cart.cartItems[i].product.addon_items.filter(addon => addon.addon_required)
        if (requiredAddons) {
          for (let x = 0, y = requiredAddons.length; x < y; x++) {
            if (
              requiredAddons &&
              requiredAddons.length > 0 &&
              (!cart.cartItems[i].activeAddons ||
                (cart.cartItems[i].activeAddons &&
                  cart.cartItems[i].activeAddons.filter(addon => addon.sku === requiredAddons[x].sku).length < 1))
            ) {
              const index = lineItems.findIndex(product => product.sku === '99999925')
              if (index > -1) {
                lineItems[index].quantity += cart.cartItems[i].quantity * requiredAddons[x].quantity
              } else {
                lineItems.push({
                  sku: '99999925',
                  quantity: parseInt(cart.cartItems[i].quantity * requiredAddons[x].quantity),
                  deliveryType: 'D',
                  required: true,
                })
              }
            }
          }
        }
        if (cart.cartItems[i].activeAddons) {
          for (let x = 0, y = cart.cartItems[i].activeAddons.length; x < y; x++) {
            const index = lineItems.findIndex(product => product.sku === cart.cartItems[i].activeAddons[x].sku)
            if (index > -1) {
              lineItems[index].quantity += cart.cartItems[i].activeAddons[x].quantity
            } else {
              lineItems.push({
                sku: cart.cartItems[i].activeAddons[x].sku,
                quantity: parseInt(cart.cartItems[i].quantity * cart.cartItems[i].activeAddons[x].quantity),
                deliveryType: cart.cartItems[i].activeAddons[x].delivery_type,
                required: true,
                title: cart.cartItems[i].activeAddons[x].title,
                category: cart.cartItems[i].activeAddons[x].category,
              })
            }
          }
        }
      }
    }
  }
  return lineItems
}

export const getProductChildren = item => {
  let children = item.product.package_skus
  let childArr = []
  if (children) {
    const region = getRegionZone().region
    children = children[region]
    if (children) {
      for (let i = 0, n = children.length; i < n; i++) {
        childArr.push({ ...children[i], quantity: parseInt(children[i].quantity) })
      }
    }
  }
  return childArr
}

export const checkManualAddress = async (order, invalidFields, billing = false) => {
  if (!order.acceptManual) {
    const address = billing ? order.billingAddress : order.shippingAddress
    const manualAddressQAS = await fetchAddressLookup(
      `${ address.address1 }, ${ address.city } ${ address.state } ${ address.zip }`
    )
    if (manualAddressQAS && manualAddressQAS.totalMatches > 0) {
      invalidFields = await addressSuggestion(order, invalidFields, manualAddressQAS, billing)
    } else if (address.zip.includes('-')) {
      const manualAddressQASNoPlus4 = await fetchAddressLookup(
        `${ address.address1 }, ${ address.city } ${ address.state } ${ address.zip.split('-')[0] }`
      )
      invalidFields = await addressSuggestion(order, invalidFields, manualAddressQASNoPlus4, billing, true)
    } else if (manualAddressQAS && manualAddressQAS.totalMatches <= 0) {
      const manualAddressQASNoMatches = await fetchAddressLookup(address.address1)
      invalidFields = await addressSuggestion(order, invalidFields, manualAddressQASNoMatches, billing, true)
    } else {
      invalidFields = ['unable to verify']
    }
  }
  return invalidFields
}

export const closeSuggestionModal = () => {
  let order = getOrder()
  order['suggestedAddress'] = null
  store.dispatch(setOrder(order))
  store.dispatch(setShippingInvalidFields([]))
}

export const addressSuggestion = async (order, invalidFields, manualAddressQAS, billing, badPlus4 = false) => {
  if (
    manualAddressQAS &&
    manualAddressQAS.results &&
    manualAddressQAS.results[0] &&
    manualAddressQAS.results[0].suggestion
  ) {
    let suggestion = manualAddressQAS.results[0].suggestion.toLowerCase()
    const address = billing ? order.billingAddress : order.shippingAddress
    if (
      !suggestion.includes(address.address1.toLowerCase()) ||
      !suggestion.includes(address.city.toLowerCase()) ||
      !suggestion.includes(address.state.toLowerCase()) ||
      (!address.zip.includes(suggestion.substr(suggestion.length - 5)) || badPlus4)
    ) {
      const suggestedAddress = await fetchAddressLookup(address.address1)
      if (
        suggestedAddress &&
        suggestedAddress.results &&
        suggestedAddress.results[0] &&
        suggestedAddress.results[0].suggestion
      ) {
        order['suggestedAddress'] = suggestedAddress.results[0].suggestion
      } else {
        order['suggestedAddress'] = null
      }
      return ['unable to verify']
    }
  } else {
    order['suggestedAddress'] = null
    return ['unable to verify']
  }
  store.dispatch(setOrder(order))
  return invalidFields
}

export const acceptAddressSuggestion = (suggestedAddress, billing = false, setBillingState = null) => {
  setAddress(suggestedAddress, billing)
  closeSuggestionModal()
  billing && setBillingState ? submitBillingAddress(setBillingState) : setCheckoutStep(null, 'shipping', 'delivery')
}

export const declineAddressSuggestion = (billing = false, setBillingState = null) => {
  let order = getOrder()
  store.dispatch(setOrder({ ...order, acceptManual: true }))
  closeSuggestionModal()
  billing && setBillingState ? submitBillingAddress(setBillingState) : setCheckoutStep(null, 'shipping', 'delivery')
}
