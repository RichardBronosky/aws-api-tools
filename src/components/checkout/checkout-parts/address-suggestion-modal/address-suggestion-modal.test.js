import React from 'react'
import renderer from 'react-test-renderer'
import ReactModal from 'react-modal'
import AddressSuggestionModal from './index'
import * as globalCheckoutHelper from '@helpers/checkout/global'
import { order1 } from '@mocks/checkoutDataMocks'

ReactModal.setAppElement('*')

describe('AddressSuggestionModal', () => {
  const testProps = {
    modalOpen: true,
    order: order1,
    suggestion: '1234 Test Address, Atlanta GA 30328',
    closeModal: jest.fn(),
    setBillingState: jest.fn(),
  }

  it('renders correctly with default props', () => {
    const tree = renderer.create(<AddressSuggestionModal { ...testProps } />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('calls acceptAddressSuggestion when user chooses to keep entered address', () => {
    const tree = renderer.create(<AddressSuggestionModal { ...testProps } />)
    globalCheckoutHelper.acceptAddressSuggestion = jest.fn()
    tree.root.findByProps({ className: 'accept-btn' }).props.onClick()
    expect(globalCheckoutHelper.acceptAddressSuggestion).toHaveBeenCalled()
  })

  it('calls declineAddressSuggestion when user chooses to keep entered address', () => {
    const tree = renderer.create(<AddressSuggestionModal { ...testProps } />)
    globalCheckoutHelper.declineAddressSuggestion = jest.fn()
    tree.root.findByProps({ className: 'keep-btn' }).props.onClick()
    expect(globalCheckoutHelper.declineAddressSuggestion).toHaveBeenCalled()
  })
})
