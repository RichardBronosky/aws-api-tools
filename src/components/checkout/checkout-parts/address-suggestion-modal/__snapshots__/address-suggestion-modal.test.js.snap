// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressSuggestionModal renders correctly with default props 1`] = `
<div>
  <div
    id="content"
  >
    <div
      className="ReactModal__Overlay ReactModal__Overlay--after-open modal-overlay"
      onClick={[Function]}
      onMouseDown={[Function]}
      style={Object {}}
    >
      <div
        aria-label="Address Suggestion Modal"
        className="ReactModal__Content ReactModal__Content--after-open address-suggestion-modal"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="dialog"
        style={Object {}}
        tabIndex="-1"
      >
        <div
          className="modal-content"
        >
          <span
            className="modal-title"
          >
            We are having trouble verifying the address as it was entered.
          </span>
          <div
            className="grid-x"
          >
            <div
              className="small-12 medium-6"
            >
              <span
                className="modal-title"
              >
                Suggested address:
              </span>
              <span
                className="address"
              >
                1234 Test Address
              </span>
              <span
                className="address"
              >
                 Atlanta GA 30328
              </span>
              <button
                aria-label="accept address suggestion"
                className="accept-btn"
                onClick={[Function]}
                tabIndex="0"
                value="accept address suggestion"
              >
                ACCEPT ADDRESS SUGGESTION
              </button>
            </div>
            <div
              className="small-12 medium-6"
            >
              <span
                className="modal-title"
              >
                Address as entered:
              </span>
              <span
                className="address"
              >
                103 seffner ave 
              </span>
              <span
                className="address"
              >
                seffner FL 33584
              </span>
              <button
                aria-label="keep address as entered"
                className="keep-btn"
                onClick={[Function]}
                tabIndex="0"
                value="keep address as entered"
              >
                KEEP ADDRESS AS ENTERED
              </button>
            </div>
          </div>
          <button
            aria-label="Close"
            className="go-back"
            onClick={[MockFunction]}
            tabIndex="0"
            value="Close"
          >
            &lt; GO BACK AND EDIT THE ADDRESS PROVIDED
          </button>
          <button
            aria-label="Close"
            className="close-modal"
            onClick={[MockFunction]}
            tabIndex="0"
            value="Close"
          >
            <img
              alt="close icon"
              className="icon close"
              src="data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 24 24'%3E%3Cpath fill='%23FFFFFF' d='M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z' /%3E%3C/svg%3E"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    data-target-tag-name="DIV"
    id="target"
  />
</div>
`;
